apps/shop/views.py

```diff
@@ -0,0 +1,321 @@
+                cart=cart
+            )
+            if new_product_count > 0:
+                cart_product_m2m.product_count = new_product_count
+                cart_product_m2m.save()
+                return Response(status=HTTP_204_NO_CONTENT)
+            else:
+                cart_product_m2m.delete()
+                return Response(status=HTTP_204_NO_CONTENT)
+
+        except CartProductM2M.DoesNotExist:
+            if new_product_count > 0:
+                CartProductM2M(
+                    **{'cart': cart, **serializer.validated_data}
+                ).save()
+
+            return Response(status=HTTP_204_NO_CONTENT)
+
+    def get(self, request, *args, **kwargs):
+        queryset = CartProductM2M.objects \
+            .filter(cart=request.user.cart) \
+            .select_related('product')
+        return Response(CartProductCountSerializer(queryset, many=True).data)
+
+
+class CatrToOrderView(APIView):
+    permission_classes = (IsAuthenticated,)
+
+    def get(self, request, *args, **kwargs):
+
+        cart = request.user.cart
+        cart_product_m2ms = CartProductM2M.objects \
+            .filter(cart=cart) \
+            .select_related('product')
+        if cart_product_m2ms.exists():
+            order = Order(user=request.user)
+            order.save()
+
+            order_product_m2ms = (
+                OrderProductM2M(
+                    order=order,
+                    product=cart_product_m2m.product,
+                    product_count=cart_product_m2m.product_count
+                )
+                for cart_product_m2m
+                in cart_product_m2ms
+            )
+            OrderProductM2M.objects.bulk_create(order_product_m2ms)
+
+            cart_product_m2ms.delete()
+
+            return Response(status=HTTP_204_NO_CONTENT)
+
+        return Response(status=HTTP_400_BAD_REQUEST)
+
+
+class ClearCartView(APIView):
+    permission_classes = (IsAuthenticated,)
+
+    def get(self, request, *args, **kwargs):
+        cart = request.user.cart
+        cart.products.clear()
+        return Response(status=HTTP_204_NO_CONTENT)
+
+
+class OrderViewSet(ListModelMixin,
+                   RetrieveModelMixin,
+                   UpdateModelMixin,
+                   DestroyModelMixin,
+                   GetSerializerClassMixin,
+                   GenericViewSet):
+    permission_classes = (IsAuthenticated,)
+
+    serializer_class = OrderSerializer
+    serializer_action_classes = {
+        'update': OrderProductCountSerializer,
+        'partial_update': OrderProductCountSerializer,
+    }
+    queryset = Order.objects.all()
+
+    def get_queryset(self):
+        return super().get_queryset() \
+            .filter(user=self.request.user) \
+            .prefetch_related('orderproductm2m_set', 'products')
+
+    def get_order_to_uptade(self):
+        order = self.get_object()
+
+        if order.status != ORDER_STATUS_INCOMPLETE:
```

Comment:
Use specific permissions class instead. This logic can be moved into `has_object_permission` method.